### ISSUE TITLE
[AuthN] Publishers want to sign up or sign in on the portal to publish their datasets

### ISSUE THREAD

**Issue by brianraymor** [chanzuckerberg/corpora-data-portal#517]
What:

1. User needs to be able to log in / log out from Data Portal. Log in button will hit `GET API_DOMAIN/v1/login` and Log out button will hit `GET API_DOMAIN/v1/logout`

1. On app initialization, makes an API call to `GET API_DOMAIN/v1/userinfo` to know whether the user is logged in or not

1. When a user logs in from cellxgene and visits Data Portal, they need to be treated as logged in at Data Portal too

1. On the other hand, if a user logs in from Data Portal and visits cellxgene, they need to be treated as logged in at cellxgene

1. We need to implement Call to Action for user to sign in / sign up when clicking on "Create Collection" button

1. If a user is logged in, follow mock design to indicate the user is logged in. If name is available, show **FULL** name, otherwise **show emoji** like `cellxgene` does

   <img width="375" alt="Screen Shot 2020-10-20 at 12 07 53 PM" src="https://user-images.githubusercontent.com/6309723/96632725-fea84100-12cc-11eb-884a-dfff93f5893a.png">

1. If a user is not logged in, follow mock design and implement log in and create account buttons
1. Check `backend/config/corpora-api.yml` for available endpoints! This [PR](https://github.com/chanzuckerberg/corpora-data-portal/pull/513/files) is the original BE AuthN implementation. 
    - Should be just `/v1/login`, `/v1/logout`, and `/v1/userinfo`. 
    - Prod example: `https://api.cellxgene.cziscience.com/dp/v1/userinfo`

1. Create a `react-query` to get `userinfo` from API

Design:
https://www.figma.com/file/g0L0CfVzRc8r6HQa4NH1wJ/Collections?node-id=0%3A1
https://www.figma.com/file/6wPUe08qSM8V9G2G0UfYOP/Consumer-Collection?node-id=36%3A585
<img width="902" alt="Screen Shot 2020-10-20 at 11 52 44 AM" src="https://user-images.githubusercontent.com/6309723/96631144-cbfd4900-12ca-11eb-9061-980f479b2952.png">

========== Background Info ==========

**Story**

_Publishers want to create an account on the portal to publish their datasets, increasing their impact through review, evaluation, and exploration by scientists with broad expertise._

**Scheduling Guidance**
We may want to schedule this _just-in-time_ after the minimum self-publishing flow is implemented. This must not be released to publishers if implemented earlier. This is the remaining _frontend_ registration issue to complete this story. 

**UX and BE Designs:**

- [Authentication](https://docs.google.com/document/d/19cC_tH7DOHIYqojdo-6T9eTzPmoMkwrAj83Ozvqth8s/edit#heading=h.t849jh3lwzxu)
- [Flow](https://github.com/chanzuckerberg/corpora-data-portal/issues/356)

**Comment by tihuan** [comment]
@bmccandless @mweiden Ah looks like BrianR created the DP AuthN FE ticket for us already!

I just added "design me" label because we need to figure out how many buttons (e.g., Sign Up / Log in) we need to create, what they look like, where to place them on the page, and whether they should look consistent with the login button on cellxgene -- @pdugan20 @colinmegill since we're talking about standardizing buttons, would love to get your input here 🙏 Thank you!

Once that done we can move the ticket to `refine me`, and assign Timmy to write out the technical details before we estimate this!

**Comment by pdugan20** [comment]
It sounded like @colinmegill wanted us to default to the standard blueprint button style? Is that still correct?

**Comment by tihuan** [comment]
@brianraymor Sorry I'm adding back the `design me` label, because I think the design step of this ticket is not locked down yet!

> I just added "design me" label because we need to figure out how many buttons (e.g., Sign Up / Log in) we need to create, what they look like, where to place them on the page, and whether they should look consistent with the login button on cellxgene -- @pdugan20 @colinmegill since we're talking about standardizing buttons, would love to get your input here 🙏 Thank you!

Currently it looks like it's pending @colinmegill 's response to @pdugan20 's question regarding the button style 😃? Thank you!

**Comment by brianraymor** [comment]
My thinking - the entire epic was already blocked by Pat's design so it was redundant to block this issue with the same design.

**Comment by tihuan** [comment]
Ahh got it! Thanks, @brianraymor ! Wasn't aware the epic was blocked by design already. Totally makes sense 👍 Removed the flag!

**Comment by pdugan20** [comment]
@tihuan I can make sure we get resolution from @colinmegill re: button styles, but my understanding is that using a blueprint style button is more important longer term and that for the version we are building now, there is probably some additional flexibility since it will change over time. colin is that correct?

**Comment by tihuan** [comment]
Hi @brianraymor ! Thanks for QAing this!

Steps:

1. visit https://cellxgene.dev.single-cell.czi.technology?auth=true&cc=true
1. You should be able to see both "Login" button and "Create Collection" button
1. If you click on "Create Collection" button, you should see the CTA to either sign up or sign in. Clicking on "Continue" button take you to the Auth0 page
1. If you click on "Login" button, it'll take you to the Auth0 page too
1. After you signup/login, you should be redirected back to DP homepage, and you'll either see your full name on the upper right corner or just the first letter of your email, if we don't have your full name
1. Now clicking on "Create Collection" will trigger the modal, but given the WIP AuthN integration last piece, you won't be able to create a collection at the moment
1. Go back to the homepage and click on any "View" button from a dataset, it should take you to its Explorer page and you should be logged in as well

That's it! Thank you!

**Comment by brianraymor** [comment]
**Maintaining Log in state between pages**

1. visit cellxgene data portal
2. sign in
3. launch cellxgene explorer
4. sign out from cellxgene explorer
5. return to the data portal page (but do not refresh) - still logged in and can create a collection
6. refresh cellxgene data portal (now I'm logged out)

when I do the reverse:
1. visit cellxgene data portal
2. do not sign in
3. launch cellxgene explorer
4. sign in
5. return to the data portal page (but do not refresh) - then I'm logged in as expected

---

**Incomplete flow for Create Collection when not signed in**

1. visit cellxgene data portal
2. Click on **Create Collection**
3. Click continue and sign in when prompted

I'm then expecting to be prompted with the Create Collection dialog, but I'm not. It's like the portal forgot why it asked me to sign in.

**Comment by tihuan** [comment]
Thanks for the detailed steps, @brianraymor !

Let me see how I can fix the first scenario as part of the AuthN integration ticket bug fix 🙏 

And for the 3rd scenario, I think we'd need another story to implement that, since it wasn't a requirement for this AuthN ticket. And it'll probably be a 3 point BE/FE full stack story, since it requires adding redirectUrl support in Auth0 that BrianM and I didn't implement. Is that something we wanna do?

Thank you!

**Comment by tihuan** [comment]
Sending this back to WIP for the bug fix!

**Comment by brianraymor** [comment]
Reviewed with @hthomas. We  _**implicitly**_ assumed scenario 3 (logged out collection view flow should result in the publisher being offered the create collection dialog upon sign in). Harley has now added an [**explicit** flow](https://www.figma.com/file/g0L0CfVzRc8r6HQa4NH1wJ/Publisher-Collection?node-id=906%3A192). I'll create an additional issue in [Publishers want to create a private collection for related dataset submissions](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell/23) to address this gap.

**Comment by tihuan** [comment]
@Bento007 A fix is out for this 😄 

https://github.com/chanzuckerberg/corpora-data-portal/pull/744

PTAL thank you!

**Comment by tihuan** [comment]
Ahh got it! Thanks so much for creating the additional issue 🙏 Look forward to looking on that!

**Comment by brianraymor** [comment]
Created [The data portal must support the intention to create a collection when the publisher is not signed in](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/corpora-data-portal/745)

**Comment by tihuan** [comment]
@brianraymor the fix has been deployed to Dev now! Scenario 1 should work now as follows:

1. visit cellxgene data portal
1. sign in
1. launch cellxgene explorer in a different tab
1. sign out from cellxgene explorer
1. return to the data portal page/tab (but do not refresh) <-- DP will now auto refresh the page to log the user out!

Thank you!
